File: 49. Group Anagrams.java
// Given an array of strings, group anagrams together.

// For example, given: ["eat", "tea", "tan", "ate", "nat", "bat"], 
// Return:

// [
//   ["ate", "eat","tea"],
//   ["nat","tan"],
//   ["bat"]
// ]
// Note: All inputs will be in lower-case.

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0) return new ArrayList<List<String>>();
        Map<String, List<String>> map = new HashMap<>();
        for (String s : strs) {
        	char [] ca = s.toCharArray();
        	Arrays.sort(ca);
        	String temp = String.valueOf(ca);
        	if (!map.containsKey(temp)) {
        		map.put(temp, new ArrayList<String>());
        	}
        	map.get(temp).add(s);
        }
        return new ArrayList<List<String>>(map.values());
    }
}

File: 49. Group Anagrams.py
# Given an array of strings, group anagrams together.

# For example, given: ["eat", "tea", "tan", "ate", "nat", "bat"], 
# Return:

# [
#   ["ate", "eat","tea"],
#   ["nat","tan"],
#   ["bat"]
# ]
import collections

class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[
        """
        if not strs:
        	return []

        # Solution 1
        dic = {}
        for s in strs:
        	t = tuple(sorted(s))
        	# or t = str(sorted(s)), 
        	# as long as the key of the dictionary be hashable it should be fine
        	dic.setdefault(t, []).append(s)
        return dic.values()

        # Solution 2
        ans = collections.defaultdict(list)
        for s in strs:
            ans[tuple(sorted(s))].append(s)
        return ans.values()
        

